File: aimDK/src/examples/src/hal/omnihand_control.cpp
#include "aimdk_msgs/msg/hand_command.hpp"
#include "aimdk_msgs/msg/hand_command_array.hpp"
#include "aimdk_msgs/msg/hand_type.hpp"
#include "aimdk_msgs/msg/message_header.hpp"
#include "rclcpp/rclcpp.hpp"
#include <chrono>
#include <vector>

/**
 * @brief Omnihand control node
 */
class OmnihandControl : public rclcpp::Node {
public:
  OmnihandControl()
      : Node("omnihand_control"), position_pairs_({
                                      {1.0, 1.0}, // Hands fully open
                                      {0.0, 0.0}, // Hands fully closed
                                      {0.5, 0.5}, // Hands half open
                                      {0.3, 0.7}, // Left closed right open
                                      {0.8, 0.2}  // Left open right closed
                                  }),
        current_index_(0) {
    publisher_ = this->create_publisher<aimdk_msgs::msg::HandCommandArray>(
        "/aima/hal/joint/hand/command", 10);

    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(20), // 50Hz
        std::bind(&OmnihandControl::publish_hand_commands, this));

    last_switch_time_ = now();
    RCLCPP_INFO(this->get_logger(), "Omnihand control node has been started!");
  }

  /**
   * @brief Publish hand commands
   */
  void publish_hand_commands() {
    // 1. Switches parameter set every 5 seconds
    auto now_time = this->now();
    if ((now_time - last_switch_time_).seconds() >= 5.0) {
      current_index_ = (current_index_ + 1) % position_pairs_.size();
      last_switch_time_ = now_time;
      RCLCPP_INFO(
          this->get_logger(),
          "Switched to parameter set %zu, left hand = %.2f, right hand = %.2f",
          current_index_, position_pairs_[current_index_].first,
          position_pairs_[current_index_].second);
    }

    auto msg = std::make_unique<aimdk_msgs::msg::HandCommandArray>();
    msg->header = aimdk_msgs::msg::MessageHeader();

    float left_position = position_pairs_[current_index_].first;
    float right_position = position_pairs_[current_index_].second;

    // Left hand 10 joints
    for (int i = 0; i < 10; ++i) {
      aimdk_msgs::msg::HandCommand left_hands;
      left_hands.name = "left_hand_joint" + std::to_string(i);
      left_hands.position = left_position;
      left_hands.velocity = 1.0;
      left_hands.acceleration = 1.0;
      left_hands.deceleration = 1.0;
      left_hands.effort = 1.0;
      msg->left_hands.push_back(left_hands);
    }
    // Right hand 10 joints
    for (int i = 0; i < 10; ++i) {
      aimdk_msgs::msg::HandCommand right_hands;
      right_hands.name = "right_hand_joint" + std::to_string(i);
      right_hands.position = right_position;
      right_hands.velocity = 1.0;
      right_hands.acceleration = 1.0;
      right_hands.deceleration = 1.0;
      right_hands.effort = 1.0;
      msg->right_hands.push_back(right_hands);
    }

    // Type 1: Dexterous hand mode
    msg->left_hand_type.value = 1;
    msg->right_hand_type.value = 1;

    publisher_->publish(std::move(msg));
  }

private:
  // Member variables
  rclcpp::Publisher<aimdk_msgs::msg::HandCommandArray>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;

  std::vector<std::pair<float, float>> position_pairs_;
  size_t current_index_;
  rclcpp::Time last_switch_time_;
};

/**
 * @brief Main function
 */
// Initializes ROS, creates node, and starts spinning
int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto omnihand_control_node = std::make_shared<OmnihandControl>();
  rclcpp::spin(omnihand_control_node);
  rclcpp::shutdown();
  return 0;
}

File: aimDK/src/py_examples/py_examples/omnihand_control.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from aimdk_msgs.msg import HandCommand, HandCommandArray, HandType, MessageHeader


class OmnihandControl(Node):
    def __init__(self):
        super().__init__('omnihand_control')

        self.num_fingers = 5
        self.joints_per_finger = 2
        self.total_joints = self.num_fingers * self.joints_per_finger  # 10

        # ===== Define 10 gesture sets directly as arrays =====
        # Each set has 20 joint params (left 10 + right 10)
        # 1 = bent, 0 = straight
        self.gesture_sequence = [
            # bend 1 finger
            [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            # bend 2 fingers
            [1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            # bend 3 fingers
            [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0],
            # bend 4 fingers
            [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0],
            # bend 5 fingers
            [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
            # release 1 finger
            [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0],
            # release 2 fingers
            [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0],
            # release 3 fingers
            [1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            # release 4 fingers
            [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
             1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            # all straight
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
             0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
        ]

        self.current_index = 0
        self.last_switch_time = self.get_clock().now()

        self.publisher = self.create_publisher(
            HandCommandArray,
            '/aima/hal/joint/hand/command',
            10
        )

        # Switch gesture every second
        self.timer = self.create_timer(
            1.0,
            self.publish_hand_commands
        )

        self.get_logger().info("Omnihand demo control node started!")

    def publish_hand_commands(self):
        now_time = self.get_clock().now()
        if (now_time - self.last_switch_time).nanoseconds / 1e9 >= 1.0:
            self.current_index = (self.current_index +
                                  1) % len(self.gesture_sequence)
            self.last_switch_time = now_time
            self.get_logger().info(
                f'Switched to gesture {self.current_index + 1}/10')

        joint_array = self.gesture_sequence[self.current_index]

        msg = HandCommandArray()
        msg.header = MessageHeader()

        # Left hand
        for i in range(self.total_joints):
            cmd = HandCommand()
            cmd.name = f"left_hand_joint{i}"
            cmd.position = joint_array[i]
            cmd.velocity = 1.0
            cmd.acceleration = 1.0
            cmd.deceleration = 1.0
            cmd.effort = 1.0
            msg.left_hands.append(cmd)

        # Right hand
        for i in range(self.total_joints):
            cmd = HandCommand()
            cmd.name = f"right_hand_joint{i}"
            cmd.position = joint_array[self.total_joints + i]
            cmd.velocity = 1.0
            cmd.acceleration = 1.0
            cmd.deceleration = 1.0
            cmd.effort = 1.0
            msg.right_hands.append(cmd)

        msg.left_hand_type = HandType()
        msg.left_hand_type.value = 1  # dexterous hand mode
        msg.right_hand_type = HandType()
        msg.right_hand_type.value = 1

        self.publisher.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = OmnihandControl()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
